components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -1112,24 +1113,86 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         traceWI(ctx, { workspaceId });
 
         const user = this.checkAndBlockUser("watchWorkspaceImageBuildLogs", undefined, { workspaceId });
-        const logCtx: LogContext = { userId: user.id, workspaceId };
-
-        const { instance, workspace } = await this.internGetCurrentWorkspaceInstance(ctx, workspaceId);
-        if (!this.client) {
+        const client = this.client;
+        if (!client) {
             return;
         }
+
+        const logCtx: LogContext = { userId: user.id, workspaceId };
+        let { instance, workspace } = await this.internGetCurrentWorkspaceInstance(ctx, workspaceId);
         if (!instance) {
             log.debug(logCtx, `No running instance for workspaceId.`);
             return;
         }
         traceWI(ctx, { instanceId: instance.id });
-        if (!workspace.imageNameResolved) {
-            log.debug(logCtx, `No imageNameResolved set for workspaceId, cannot watch logs.`);
-            return;
-        }
         const teamMembers = await this.getTeamMembersByProject(workspace.projectId);
         await this.guardAccess({ kind: "workspaceInstance", subject: instance, workspace, teamMembers }, "get");
-        if (!this.client) {
+
+        // wait for up to 20s for imageBuildLogInfo to appear due to:
```

Comment:
What's the behaviour of those things don't happen within 20s?
Can we get away without the time-based behaviour?

Reply:
> What's the behaviour of those things don't happen within 20s?

The (browser) client re-trys (if this is still wanted).

> Can we get away without the time-based behaviour?

Only without db-sync, and only if we had a way to know the definite state of an `imagebuild`. We either:
 - enhance the field `workspace.imageBuildInfo` to also contain info about the state of the build (pending, pulling, running, stopped)
 - or we re-visit our decision to surface `imagebuild` on the application side (e..g, DB), so we could read that state easily

Reply:
@csweichel :point_up: This would also the lifecycle-mismatch of `imageBuildInfo`: The current change stores it in `Workspace` because it is convenient. And this is correct as long as we never have to re-build an image. But once we have to - think of docker image cleanup, registry data loss, SH user re-sets a PV, etc. - we're updating the workspace again, which is kind of weird.

Having a own entity for this akin to prebuilds would really make sense. Conceptually this could also help with explaining image builds better to users by providing more transparency in the UI.

I'm not 100% settled we want that (and whether/when the effort is justified), just this PR got me thinking in that direction again.

Reply:
Another argument: The regionality split between imagebuilds and workspaces. With this PR we potentially override `imageBuildInfo` from different regions when building images with different `image-builder-mk3` instances.

Reply:
what if we stored the `imageBuildInfo` on the instance instead of the workspace? That's much cluster to the actual lifecycle of the build, as every instance can potentially require an image build.

Reply:
Yes, that would fix problem 2 and 3 (the lifecycle and regionality "issues"). :+1: Still we'll have the "not enough state" information. I will go with that approach for this PR.

We can think about how to surface image build later. But IMO this is sth we should solve, also to help make image builds more transparent to the user. For instance by offering more UI that explains:
 - when an image build happened
 - what triggered it
 - why it failed
 - how to fix it

BTW this is connected to "prebuild reliability" /cc @jldec